File: mini-extractor-go/extractor/extractor.go
package extractor

import (
	"context"
	"fmt"

	"github.com/ChefBingbong/mini-extractor-go/lib"
	"github.com/ChefBingbong/viem-go/client"
)

// Config holds the configuration for the Extractor.
type Config struct {
	Client      *client.PublicClient
	FactoriesV2 []FactoryV2
	CacheDir    string
	LogType     LogFilterType
	LogDepth    int
	Debug       bool
}

// Extractor is the top-level orchestrator that ties together the
// LogFilter, TokenManager, and UniV2Extractor.
type Extractor struct {
	Client       *client.PublicClient
	ExtractorV2  *UniV2Extractor
	TokenManager *TokenManager
	LogFilter    *LogFilter2
	Config       Config
}

// NewExtractor creates a new Extractor from the given config.
func NewExtractor(cfg Config) *Extractor {
	chainID := 0
	if cfg.Client.Chain() != nil {
		chainID = int(cfg.Client.Chain().ID)
	}

	tokenManager := NewTokenManager(
		cfg.Client,
		chainID,
		cfg.CacheDir,
		fmt.Sprintf("tokens-%d", chainID),
	)

	logFilter := NewLogFilter2(
		cfg.Client,
		cfg.LogDepth,
		cfg.LogType,
		cfg.Debug,
	)

	extractorV2 := NewUniV2Extractor(
		cfg.Client,
		cfg.FactoriesV2,
		cfg.CacheDir,
		logFilter,
		tokenManager,
		chainID,
	)

	return &Extractor{
		Client:       cfg.Client,
		ExtractorV2:  extractorV2,
		TokenManager: tokenManager,
		LogFilter:    logFilter,
		Config:       cfg,
	}
}

// Start initializes the log filter and starts the V2 extractor.
func (e *Extractor) Start(ctx context.Context) error {
	e.LogFilter.Start()
	if err := e.ExtractorV2.Start(ctx); err != nil {
		return err
	}
	lib.Info("Extractor started successfully")
	return nil
}

// GetPools returns all known pool states.
func (e *Extractor) GetPools() []PoolState {
	return e.ExtractorV2.GetPools()
}

// GetPoolMap returns the pool map.
func (e *Extractor) GetPoolMap() map[string]*PoolState {
	return e.ExtractorV2.GetPoolMap()
}

// IsStarted returns true if the extractor has completed initial startup.
func (e *Extractor) IsStarted() bool {
	return e.ExtractorV2.IsStarted()
}

// IsSyncing returns true if background factory sync is in progress.
func (e *Extractor) IsSyncing() bool {
	return e.ExtractorV2.IsSyncing()
}

File: mini-extractor-ts/extractor/Extractor.ts
import EventEmitter from 'node:events'
import type { PublicClient } from 'viem'
import { logger } from '../lib/logger.js'
import { LogFilter2, LogFilterType } from './LogFilter2.js'
import { TokenManager } from './TokenManager.js'
import { UniV2Extractor } from './UniV2Extractor.js'
import type { FactoryV2, PoolState } from './UniV2Types.js'

export class BlockEmitter extends EventEmitter {}

export type ExtractorConfig = {
  client: PublicClient
  factoriesV2: FactoryV2[]
  cacheDir: string
  logType?: LogFilterType
  logDepth: number
  maxPools?: number // 0 or undefined = unlimited; >0 = stop background sync after this many pools
  logging?: boolean
  debug?: boolean
}

export class Extractor {
  client: PublicClient
  extractorV2: UniV2Extractor
  tokenManager: TokenManager
  readonly logFilter: LogFilter2
  blockEmitter: EventEmitter
  config: ExtractorConfig

  constructor(args: ExtractorConfig) {
    this.config = args
    this.client = args.client
    this.blockEmitter = new BlockEmitter()

    const chainId = args.client.chain?.id as number

    this.tokenManager = new TokenManager(
      args.client,
      args.cacheDir,
      `tokens-${chainId}`,
    )

    this.logFilter = new LogFilter2(
      this.client,
      args.logDepth,
      args.logType ?? LogFilterType.OneCall,
      this.blockEmitter,
      args.logging,
    )

    this.extractorV2 = new UniV2Extractor(
      this.client,
      args.factoriesV2,
      args.cacheDir,
      this.logFilter,
      this.tokenManager,
      args.maxPools ?? 0,
    )
  }

  async start() {
    this.logFilter.start()
    await this.extractorV2.start()
    logger.info('Extractor started successfully')
  }

  getPools(): PoolState[] {
    return this.extractorV2.getPools()
  }

  getPoolMap(): Map<string, PoolState> {
    return this.extractorV2.getPoolMap()
  }

  isStarted(): boolean {
    return this.extractorV2.isStarted()
  }

  isSyncing(): boolean {
    return this.extractorV2.isSyncing()
  }
}
